crates/cacache/src/put.rs

```diff
@@ -47,6 +47,165 @@ where
+    let mut dest = SSRIStream {
+        inner: std::io::Cursor::new(Vec::with_capacity(1024 * 1024 * 10)),
+        builder: ssri::IntegrityOpts::new().algorithm(ssri::Algorithm::Sha256)
+    };
+    let mut pb = cache.as_ref().to_owned();
+
+    async_std::task::spawn(async move {
+        let mut offsets = std::collections::BTreeMap::new();
+        let mut offset = 0;
+        while let Some(entry) = entry_stream.next().await {
+            let path = entry.path();
+            if path.is_err() {
+                continue
+            }
+            let path = path.unwrap();
+            let size = entry.size();
+            let mode = entry.mode().unwrap_or(0o644);
+
+            offset += dest.write(size.to_be_bytes().as_ref()).to_internal()?;
+            let mut encoded = snap::write::FrameEncoder::new(dest);
+            let mut entry = SSRIStream {
+                inner: entry,
+                builder: ssri::IntegrityOpts::new().algorithm(ssri::Algorithm::Sha256)
+            };
+
+            offset += std::io::copy(&mut entry, &mut encoded).to_internal()? as usize;
+            let (sri, _) = entry.into_inner();
+            dest = encoded.into_inner().to_internal()?;
+
+            // Would be nice if SSRI provided "into_bytes()" possibly?
+            let hexed = sri.to_hex().1;
+            offsets.insert(hex::decode(hexed).to_internal()?, offset);
+            entry_hash.insert(path, (sri, size, mode));
+        }
+
+        std::mem::drop(entry_stream);
+
+        let index_object = bincode::serialize(&entry_hash).to_internal()?;
+        let index_size = index_object.len();
+        dest.write_all(index_size.to_be_bytes().as_ref()).to_internal()?;
+        let mut encoded = snap::write::FrameEncoder::new(dest);
+        let mut entry = SSRIStream {
+            inner: std::io::Cursor::new(index_object),
+            builder: ssri::IntegrityOpts::new().algorithm(ssri::Algorithm::Sha256)
+        };
+
+        std::io::copy(&mut entry, &mut encoded).to_internal()?;
+        let (index_sri, _) = entry.into_inner();
+
+        dest = encoded.into_inner().to_internal()?;
+        let (packfile_sri, mut output_cursor) = dest.into_inner();
+
+        let offsets = offsets.into_iter().collect::<Vec<_>>();
+        let mut fanout = [0u8; 2048];
+        let mut last_index = 0;
+
+        // the fanout is a table of 256 entries, keyed positionally.
+        // the index of this fanout represents the first byte of a given
+        // integrity hash. it is mapped to an offset in a stream of hashes+offsets
+        // that follows the packfile.
+        for i in 0..255 {
+            'inner:
+            for (integrity_bytes, offset) in &offsets[last_index..] {
+                if integrity_bytes[0] != i {
+                    break 'inner;
+                }
+
+                last_index += 1;
+            }
+
+            &mut fanout[i as usize * 8..].copy_from_slice((last_index as u64).to_be_bytes().as_ref());
+        }
```

Comment:
You end up with 256 entries pointing at indices in the list of offsets. Each entry points to the first object that does not match the first byte of the index. So, assuming you had objects with integrities of `001`, `002`, and `01f`, `fanout[0]` would point at `2` – the index of `01f`.

The other neat property here is that you can get a count of the total number of objects in the packfile by looking at `fanout[0xff]`. This lets us quickly hop into the object-offsets-sorted-by-integrity, then do a binary search for the object we're looking for.